### DISCUSSION TITLE
[30 Giugno 2022] Cache

### DISCUSSION THREAD

**Discussion by Elia-Belli** [sapienzastudentsnetwork/architettura-degli-elaboratori#61] [Cache (M-Z)]
![Esame](https://github.com/sapienzastudentsnetwork/archelab2223/assets/107944476/5d6b6b3f-1d04-4a87-b049-dcf7eba71fa1)

**Comment by daromacarlo** [comment]
![Cattura4](https://github.com/sapienzastudentsnetwork/archelab2223/assets/135441131/f6198413-cafc-4ce6-ac42-0c1591f0831f)
![Cattura3](https://github.com/sapienzastudentsnetwork/archelab2223/assets/135441131/437d073c-5b21-40fa-83a4-89f03cb66bfe)
![Cattura2](https://github.com/sapienzastudentsnetwork/archelab2223/assets/135441131/2ba29b94-e20c-48ed-a9c1-49df0527d91f)
![Cattura1](https://github.com/sapienzastudentsnetwork/archelab2223/assets/135441131/b26529d9-c2a2-420c-bd60-891965f009b6)
correzioni in caso di errore sono gradite :))
la cache fully associative nella prima immagine l'ho usata per capire se ci fossero miss di capacity

**Comment by RaulComaz** [comment]
l'esercizio 3 il punto 3 è sbagliato perchè fai tempo medio x (CPI * Periodo Clock)
la sol corretta dovrebbe essere N.I = 2202 / 2 = 91, 75

**Comment by daromacarlo** [comment]
@GrenzGaenger450 non dovrebbe essere tempomedio/(CPI*CLK) ?

**Comment by FeddyLix17** [comment]
## 1) Accessi in memoria

dopo essersi ricavati le formule usando un approccio grafico

![image](https://github.com/sapienzastudentsnetwork/archelab2223/assets/66442720/935c09a9-450c-444a-bf79-a7bbbae89108)

si riempie la tabella ottenendo

![image](https://github.com/sapienzastudentsnetwork/archelab2223/assets/66442720/197e74fe-9381-4d70-bb1c-ac11d7a8a5d0)

## 2) Dimensioni delle cache

considerando che ogni linea della cache è formata

- dal bit di validità

- dal *Least Recently Used* bit (solo L1)

- dal tag

- dal blocco stesso

si ottiene, rispettivamente per le 2 cache
 
- $\Large \text{L1}$
  
   il numero delle linee è dato dal prodotto tra le vie e i set, da cui

$$
    \Large  (1  + 1 + 27 + 2 * 4 * 8) * (2 * 4) \implies (29 + 64) * 8 \implies  93 * 8 \implies 744\ \text{bit}
$$

- $\Large \text{L2}$

$$
    \Large  (1 + 24 + 16 * 4 * 8) * 4 \implies (25 + 512) * 4 \implies  537 * 4 \implies 2148\ \text{bit}
$$

## 3) Tempi della sequenza di accessi

### (a) Tempo totale sequenza di accessi

rielencandoli brevemente dalla foto superiore, si hanno

- 5 accessi in memoria, con tempo $\Large 400 * 5 = 2000\ \text{ns}$

- 2 accessi in L1, con tempo  $\Large 1 * 2 = 2\ \text{ns}$

- 5 accessi in L2, con tempo  $\Large 40 * 5 = 200\ \text{ns}$

$$
    \Large T_{tot} = 2000 + 2 + 200 = 2202\ \text{ns}
$$

### (b) Tempo medio sequenza di accessi

$$
    \Large T_{med} = \frac{\text{T.\ Totale}}{\text{Num.\ accessi}} = \frac{2202}{12} = 183.50\ \text{ns}
$$

### (c) Istruzioni svolte nel tempo medio

$$
    \Large \text{Istruzioni\ eseguite}  = \frac{\text{T.\ medio} *\ \text{Frequenza}}{\text{Clock\ per\ Istruction}} = \frac{183.50 * 2}{4} = 91.75
$$

## 4) Word offset indirizzo 8000 Cache L1

$$
    \Large \text{Word Offset} = (\text{Address} \gg \text{Byte Offset})\ \\%\ 2^{\text{bit Word Offset}}
$$

$$
    \Large  (8000 \gg 2)\ \\%\ 2^{1} \implies 2000\ \\%\ 2 \implies 0
$$

## 5) Cache Hit & Page Table Miss

Impossibile, in quanto il dato non possa essere nella cache se la pagina non sia attualmente caricata
